t/SocketWithLogging.hpp

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <boost/asio/buffers_iterator.hpp>
+#include <boost/system/error_code.hpp>
+#include <iostream>
+#include <utility>
+
+namespace bredis {
+namespace test {
+
+template <typename NextLayer> class SocketWithLogging {
+    NextLayer stream_;
+
+  public:
+    template <typename... Args>
+    SocketWithLogging(Args &&... args) : stream_(std::forward<Args>(args)...) {}
+
+    void cancel() {
+        std::cout << "{cancel}" << std::endl;
+        stream_.cancel();
+    }
+
+    template <typename BufferSequence>
+    void dump(const char *prefix, const BufferSequence &buffers) {
+        using Iterator = boost::asio::buffers_iterator<BufferSequence, char>;
+
+        Iterator it = Iterator::begin(buffers);
+        Iterator end = Iterator::end(buffers);
+        auto size = std::distance(it, end);
+
+        std::cout << "{" << prefix << " " << size << " bytes}[";
+        for (; it != end; it++) {
+            std::cout << *it;
+        }
+        std::cout << "]" << std::endl;
+    }
+
+    template <typename MutableBufferSequence, typename ReadHandler>
+    inline void async_read_some(const MutableBufferSequence &buffers,
+                                ReadHandler handler) {
+        stream_.async_read_some(buffers, handler);
+        dump("async_read_some", buffers);
+    }
+
+    template <typename ConstBufferSequence, typename WriteHandler>
+    inline void async_write_some(const ConstBufferSequence &buffers,
+                                 WriteHandler handler) {
+        dump("async_write_some", buffers);
+        stream_.async_write_some(buffers, handler);
+    }
+
+    template <typename ConstBufferSequence>
+    std::size_t write_some(const ConstBufferSequence &buffers) {
+        dump("write_some", buffers);
+        return stream_.write_some(buffers);
+    }
+
+    template <typename ConstBufferSequence>
+    std::size_t write_some(const ConstBufferSequence &buffers,
+                           boost::system::error_code &ec) {
+        dump("write_some", buffers);
+        return stream_.write_some(buffers, ec);
+    }
+
+    template <typename MutableBufferSequence>
+    std::size_t read_some(const MutableBufferSequence &buffers) {
+        auto result = stream_.read_some(buffers);
+        dump("read_some", buffers);
```

Comment:
You have a bug here. `dump` is going to print the entire buffer, but what if `read_some` did not fill the buffer? You need to trim the buffer down to `result` bytes. You could do this two ways:

1. Pass the `result` value into `dump` and account for it in your printing loop
2. Use a buffer sequence adaptor which reduces the length. An example of this is here:
https://github.com/vinniefalco/Beast/blob/master/include/beast/core/prepare_buffers.hpp#L23

To use it your code would look like this
```
    dump("read_some", beast::prepare_buffers(result, buffers));
```

If you go the route of **2** you should make it a `detail` class and not a public interface.